### <center>Algoritmos e Estruturas de Dados</center>

#### <center>Números reais</center>


**Prof.**: João Paulo Orlando

**Curso**: Técnico em Informática Integrado ao Ensino Médio

Os números de ponto flutuante são representados no hardware do computador como frações de base 2 (binárias). Por exemplo, a fração decimal 0.125 tem valor 1/10 + 2/100 + 5/1000.

Infelizmente, muitas frações decimais não podem ser representadas precisamente como frações binárias. O resultado é que, em geral, os números decimais de ponto flutuante que você digita acabam sendo armazenados de forma apenas aproximada, na forma de números binários de ponto flutuante.

O problema é mais fácil de entender primeiro em base 10. Considere a fração 1/3. Podemos representá-la aproximadamente como uma fração base 10:

`0.3`

ou melhor,

`0.33`

ou melhor,

`0.333`

e assim por diante. Não importa quantos dígitos você está disposto a escrever, o resultado nunca será exatamente `1/3`, mas será uma aproximação de cada vez melhor de `1/3`.

Da mesma forma, não importa quantos dígitos de base 2 estejas disposto a usar, o valor decimal 0.1 não pode ser representado exatamente como uma fração de base 2. No sistema de base 2, `1/10` é uma fração binária que se repete infinitamente:

`0.0001100110011001100110011001100110011001100110011...`

Se parares em qualquer número finito de bits, obterás uma aproximação. Hoje em dia, na maioria dos computadores, as casas decimais são aproximados usando uma fração binária onde o numerado utiliza os primeiros 53 bits iniciando no bit mais significativo e tendo como denominador uma potência de dois. No caso de `1/10`, a fração binária seria `602879701896397 / 2 ** 55` o que chega bem perto, mas mesmo assim, não é igual ao valor original de `1/10`.

É fácil esquecer que o valor armazenado é uma aproximação da fração decimal original, devido à forma como os floats são exibidos no interpretador interativo. O Python exibe apenas uma aproximação decimal do verdadeiro valor decimal da aproximação binária armazenada pela máquina. Se o Python exibisse o verdadeiro valor decimal da aproximação binária que representa o decimal 0.1, seria necessário mostrar:

`
0.1
0.1000000000000000055511151231257827021181583404541015625`

Contém muito mais dígitos do que é o esperado e utilizado pela grande maioria dos desenvolvedores, portanto, o Python limita o número de dígitos exibidos, apresentando um valor arredondado, ao invés de mostrar todas as casas decimais:

`
1 / 10
0.1`

Lembre-se, mesmo que o resultado impresso seja o valor exato de 1/10, o valor que verdadeiramente estará armazenado será uma fração binária representável que mais se aproxima.

In [19]:
print(1/3)

0.3333333333333333


In [20]:
print(f"{1/3:.100f}")

0.3333333333333333148296162562473909929394721984863281250000000000000000000000000000000000000000000000


In [39]:
# Importa a biblioteca com funções prontas. Estamos usando a função que calcula o cosseno com math.cos()
import math

x = 0.5

print(f'cos({x}) = {math.cos(x)}')

cos(0.5) = 0.8775825618903728


Podemos modificar o número de casas decimais da mensagem de saída, repare na diferença para o anterior. Adicionamos ```:.1f``` e ```:.3f``` entre as chaves.

In [28]:
import math

x = 0.5

print(f'cos({x}) = {math.cos(x):.1f}')
print(f'cos({x}) = {math.cos(x):.3f}')

cos(0.5) = 0.9
cos(0.5) = 0.878


Por enquanto, o que vamos precisar saber, é somente formatar as casas decimais da saída. Vários exercícios irão solicitar isso de vocês.

Entretanto, a formatação possui diversos recursos bem interessantes. Para saber mais, consulte a [documentação](https://docs.python.org/3/library/string.html#format-specification-mini-language).

In [1]:
# Faça um programa que leia 2 valores float (números reais) e exiba as 4 operações básicas.
#    mostrando os valores sempre com 4 casas decimais

valor_a = float(input())
valor_b = float(input())

soma = valor_a + valor_b
subtracao = valor_a - valor_b
multiplicacao = valor_a * valor_b
divisao = valor_a / valor_b

print(f"SOMA = {soma:.4f}")
print(f"SUBTRAÇÃO = {subtracao:.4f}")
print(f"MULTIPLICAÇÃO = {multiplicacao:.4f}")
print(f"DIVISÃO = {divisao:.4f}")


1.55
3.33
SOMA = 4.8800
SUBTRAÇÃO = -1.7800
MULTIPLICAÇÃO = 5.1615
DIVISÃO = 0.4655


In [3]:
# Faça um programa que leia 3 notas numerada de 0.0 até 10.0, 
#    calcule e exiba a média com 2 casas decimais.

n1 = float(input())
n2 = float(input())
n3 = float(input())

media = (n1+n2+n3)/3

print(f"Sua média foi {media:.2f}")


10
9
4
Sua média foi 7.67


In [4]:
#Mesma coisa do programa anterior, mas agora calcule a média ponderada, 
#    onde as notas tem os seguintes pesos: 25, 30 e 45.

n1 = float(input())
n2 = float(input())
n3 = float(input())

media = (n1*25+n2*30+n3*45)/100

print(f"Sua média foi {media:.2f}")

10
10
0
Sua média foi 5.50


In [5]:
# Faça um programa em python que calcule a área de um triângulo
#   dado a base e a altura em números float. 
#   Imprima a área com 3 casas decimais.

###   A = b*h/2    ###


base = float(input())
altura = float(input())

area = base*altura/2
print(f"Área: {area:.3f}")


10
2.5
Área: 12.500


In [ ]:
# Faça um programa em python que leia um grau em Celsius e converta ele para Fahrenheit.
#     Exiba o valores com duas casas decimais.

###    F = (C * 9/5) + 32  ###

In [15]:
# Faça um programa que calcule o IMC de uma pessoa.
# Leia o Nome, Peso e Altura da pessoa.
# Aplique a fórmula:        IMC = peso/(altura**2)
# Exiba o resultado com 1 casa decimal.



Vamos analisar uma situação peculiar de alguns problemas que vocês irão resolver de agora em diante. Observe as entradas do problema 1013.

<img src="./Imagens/05-01.png">

O método `input()` lê uma linha inteira de informações, independente do que está escrito lá.

Vamos simular a entrada desse problema:

In [1]:
entrada = input()
print(f"""A entrada foi '{entrada}'""")

7 14 106
A entrada foi '7 14 106'


O que é a informação '7 14 106'?

É somente um texto, precisamos separar esse texto, exatamente em 3 partes. 

Para isso, existe um método chamado `split()` que corta o texto pelos espaços em branco. Esse método será aplicado após a execução do `input()`.

In [3]:
valorA, valorB, valorC = input().split()
print(f"""Agora temos 3 valores: '{valorA}', '{valorB}' e '{valorC}'""")


7 14 106
Agora temos 3 valores: '7', '14' e '106'


Entretanto, temos mais problema, precisamos converter cada um dos elementos para seu tipo adequado, pois eles são somente texto e não números.

In [4]:
print(type(valorA))

<class 'str'>


In [5]:
valorA = int(valorA)
print(type(valorA))

<class 'int'>


Então a leitura desses 3 valores de entrada na mesma linha poderia ficar algo assim.


In [7]:
valorA, valorB, valorC = input().split()
valorA = int(valorA)
valorB = int(valorB)
valorC = int(valorC)

#Podendo continuar o desenvolvimento de assim por diante.


7 14 106


Agora podem começar a fazer a **Lista 3 - Números reais**.

**Dica**: A quantidade de linhas de entrada, será a quantidade de `input()` que deverá ser feito no seu programa.
